app/src/main/kotlin/com/github/gotify/api/CertUtils.kt

```diff
@@ -43,24 +49,36 @@ internal object CertUtils {
     fun applySslSettings(builder: OkHttpClient.Builder, settings: SSLSettings) {
         // Modified from ApiClient.applySslSettings in the client package.
         try {
+            var customManagers = false
+            var trustManagers: Array<TrustManager>? = null
+            var keyManagers: Array<KeyManager>? = null
+            if (settings.caCertPath != null) {
+                val tempTrustManagers = certToTrustManager(settings.caCertPath)
+                if (tempTrustManagers.isNotEmpty()) {
+                    trustManagers = tempTrustManagers
+                    customManagers = true
+                }
+            }
+            if (settings.clientCertPath != null) {
+                val tempKeyManagers = certToKeyManager(
+                    settings.clientCertPath,
+                    settings.clientCertPassword
+                )
+                if (tempKeyManagers.isNotEmpty()) {
+                    keyManagers = tempKeyManagers
+                }
+            }
             if (!settings.validateSSL) {
-                val context = SSLContext.getInstance("TLS")
-                context.init(arrayOf(), arrayOf<TrustManager>(trustAll), SecureRandom())
-                builder.sslSocketFactory(context.socketFactory, trustAll)
+                trustManagers = arrayOf(trustAll)
                 builder.hostnameVerifier { _, _ -> true }
-                return
             }
-            val cert = settings.cert
-            if (cert != null) {
-                val trustManagers = certToTrustManager(cert)
-                if (trustManagers.isNotEmpty()) {
-                    val context = SSLContext.getInstance("TLS")
-                    context.init(arrayOf(), trustManagers, SecureRandom())
-                    builder.sslSocketFactory(
-                        context.socketFactory,
-                        trustManagers[0] as X509TrustManager
-                    )
-                }
+            if (customManagers || !settings.validateSSL) {
```

Comment:
Does this mean we don't initialize the sslSocketFactory if we have a clientCertPath specified without (caCertPath || !validateSSL). Couldn't a user a have a valid ca certificate from a CA and then only want to use a client certificate?

Reply:
After a quick peek into the code I suppose I just forget the `customManagers = true` for the key managers, you're right.  
That's why I will test the different use cases in the next days

Reply:
Actually just tested the use cases. The client-cert case was broken, everything else seems to work.  
Here's what I tested for each use case:

1. Login
2. Fetching messages
3. Loading images in message (images from the internet and behind local reverse proxy with client cert auth)
4. Websocket messages